src/compact_live.rs

```diff
@@ -0,0 +1,460 @@
+            )));
+        }
+
+        let uids = self.fetch_all_entry_table_uids().await?;
+
+        // The esperance of the number of UIDs to draw in order to recompact the
+        // full Entry Table after `num_reindexing_before_full_set` operations:
+        //
+        // n_compact = (n / T) * (log(n) + gamma) + O(1/(T * n))
+        //
+        // with `n = entry_table_length`, `T = num_reindexing_before_full_set`
+        // and `gamma` is the Euler's constant.
+        //
+        // See the [coupon collector's problem](https://wikipedia.org/wiki/Coupon_collector's_problem).
+        let entry_table_length = uids.len() as f64;
+        let n_compact = ((entry_table_length * (entry_table_length.log2() + GAMMA))
+            / f64::from(num_reindexing_before_full_set))
+        .ceil() as usize;
+
+        // The number of compacted UIDs should leave enough unused UIDs for the noise.
+        if (n_compact as f64  * (1f64 + Self::NOISE_RATIO)) > uids.len() as f64 {
+            return Err(Error::CryptoError(format!(
+                "Number of Entry Table UIDs to compact ({n_compact}) should not be greater than {}",
+                uids.len() as f64 / (1f64 + Self::NOISE_RATIO)
+            )));
+        }
+
+        // Upper bounds are used for capacity in order to avoid reallocations:
+        // - the number of noise UIDs is lower than `n_compact`;
+        let mut noise = HashSet::with_capacity(n_compact);
+        // - the total number of UIDs (target set + noise set) is lower than `2*n_compact`.
+        let mut noisy_uids_to_compact = Vec::with_capacity(2 * n_compact);
+
+        // Needed because `uids` is moved in the loop condition.
+        let n_uids = uids.len();
+        let n_noise_candidates = n_uids - n_compact;
+        let n_noise = (Self::NOISE_RATIO * n_compact as f64) as usize;
+
+        for uid in uids {
+            let tmp = rng.next_u32() as usize;
+            if tmp % n_uids < n_compact {
+                // Randomly select ~ `n_compact` UIDs.
+                noisy_uids_to_compact.push(uid);
+            } else if tmp % n_noise_candidates < n_noise {
+                // Randomly select ~ `NOISE_RATIO * n_compact` UIDs.
+                noise.insert(uid);
+            }
+        }
+
+        Ok((noisy_uids_to_compact, noise))
+    }
+
+    /// Fetch all useful information for the compact from the Chain Table:
+    /// - the map of Entry Table UIDs to Chain Table UIDs
+    /// - the map of Entry Table UIDs to indexed values
+    ///
+    /// # Paramters
+    ///
+    /// - `k_value`                 : DEM key used to decrypt the Entry Table
+    /// - `encrypted_entry_table`   : encrypted Entry Table
+    async fn fetch_chain_data(
+        &self,
+        k_value: &DemScheme::Key,
+        encrypted_entry_table: &EncryptedTable<UID_LENGTH>,
+    ) -> Result<
+        (
+            HashMap<Uid<UID_LENGTH>, Vec<Uid<UID_LENGTH>>>,
+            HashMap<Uid<UID_LENGTH>, HashSet<IndexedValue>>,
+        ),
```

Comment:
Could you create a small struct instead of a 4 elements tuple?

Reply:
Done in 899799d